File: src/main/java/com/durhack/sharpshot/Coordinate.java
package com.durhack.sharpshot;


import java.util.Objects;

/**
 * This is an immutable class.
 */
public class Coordinate {
    final private int x;
    final private int y;

    public Coordinate(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public Coordinate plus(Direction direction){
        int newX = x + direction.getDeltaX();
        int newY = y + direction.getDeltaY();
        return new Coordinate(newX, newY);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Coordinate that = (Coordinate) o;
        return x == that.x &&
                y == that.y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    public Coordinate wrap(int width, int height) {
        return new Coordinate((x+width) % width, (y+height) % height);
    }
}

File: src/main/java/com/durhack/sharpshot/Coordinate.kt
package com.durhack.sharpshot


import java.util.*

/**
 * This is an immutable class.
 */
class Coordinate(val x: Int, val y: Int) {

    operator fun plus(direction: Direction): Coordinate {
        val newX = x + direction.deltaX
        val newY = y + direction.deltaY
        return Coordinate(newX, newY)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as Coordinate?
        return x == that!!.x && y == that.y
    }

    override fun hashCode(): Int {
        return Objects.hash(x, y)
    }

    fun wrap(width: Int, height: Int): Coordinate {
        return Coordinate((x + width) % width, (y + height) % height)
    }
}
